### ISSUE TITLE
[MNG-2802] Concurrent-safe access to local Maven repository

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#4925]
**[Stepan Roh](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=src)** opened **[MNG-2802](https://issues.apache.org/jira/browse/MNG-2802?redirect=false)** and commented

It seems that access to local Maven repository is not concurrent-safe that is multiple Mavens running in parallel may damage contents of local Maven repository. It would be a nice improvement, because sharing of local repository will lower the need for contacting any other repository. I know that Maven proxy can be used, but that adds another layer which may unnecessarily stress the machine it runs on.

**NOTE:** http://takari.io/book/30-team-maven.html#concurrent-safe-local-repository is a Takari Extension for Apache Maven that introduces a concurrent-safe local-repository.

---

**Issue Links:**
- [MNG-6281](https://issues.apache.org/jira/browse/MNG-6281) ArrayIndexOutOfBoundsException caused by pom.xml with invalid/duplicate XML
 (_**"is a parent of"**_)
- [MNG-4706](https://issues.apache.org/jira/browse/MNG-4706) Multithreaded building can create bad files for downloaded artifacts in local repository
 (_**"is a parent of"**_)
- [MNG-5307](https://issues.apache.org/jira/browse/MNG-5307) NPE during resolution of dependencies - parallel mode
 (_**"is a parent of"**_)
- [MNG-6604](https://issues.apache.org/jira/browse/MNG-6604) Intermittent failures while downloading GAVs from Nexus
 (_**"is a parent of"**_)
- [MNG-5315](https://issues.apache.org/jira/browse/MNG-5315) Artifact resolution sporadically fails in parallel builds
 (_**"is a parent of"**_)
- [MNG-6216](https://issues.apache.org/jira/browse/MNG-6216) ArrayIndexOutOfBoundsException when parsing POM
 (_**"fixes"**_)
- [MRESOLVER-131](https://issues.apache.org/jira/browse/MRESOLVER-131) Introduce a Redisson-based SyncContextFactory
 (_**"is fixed by"**_)
- [MDEP-442](https://issues.apache.org/jira/browse/MDEP-442) Failed to access file due to locked access when using more than one Maven worker thread
 (_**"is duplicated by"**_)
- [MDEP-518](https://issues.apache.org/jira/browse/MDEP-518) Dependency resolution fails in threaded builds
 (_**"is duplicated by"**_)
- [MNG-4071](https://issues.apache.org/jira/browse/MNG-4071) Support additional read-only local repositories
 (_**"is duplicated by"**_)
- [MRESOLVER-123](https://issues.apache.org/jira/browse/MRESOLVER-123) Provide a global locking sync context by default

- [HADOOP-11998](https://issues.apache.org/jira/browse/HADOOP-11998) Jenkins has race conditions when reading the maven repo

- [MNG-4706](https://issues.apache.org/jira/browse/MNG-4706) Multithreaded building can create bad files for downloaded artifacts in local repository

- [MNG-6996](https://issues.apache.org/jira/browse/MNG-6996) Upgrade Maven Resolver to 1.6.1
 (_**"requires"**_)
- [MDEP-657](https://issues.apache.org/jira/browse/MDEP-657) Maven - Failing when two modules have same dependency and both trying to download the dependency at the same time when Parallel flag is enabled.
 (_**"is depended upon by"**_)
- [CONTINUUM-265](https://issues.apache.org/jira/browse/CONTINUUM-265) Concurrent builds
 (_**"is depended upon by"**_)
- [MNG-3004](https://issues.apache.org/jira/browse/MNG-3004) Allow build lifecycle to execute projects in parallel
 (_**"is depended upon by"**_)
- [MNG-3204](https://issues.apache.org/jira/browse/MNG-3204) Parallel dependency fetching
 (_**"is depended upon by"**_)

**Remote Links:**
- [GitHub Pull Request #49
](https://github.com/apache/maven-wagon/pull/49)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.2](https://github.com/apache/maven/milestone/103?closed=1)

65 votes, 84 watchers

**Comment by jira-importer** [comment]
**[Brian E. Fox (imported)](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf@infinity.nu)** commented

John, you were recently fixing this in 2.1 i believe

**Comment by jira-importer** [comment]
**[John Dennis Casey](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jdcasey)** commented

To go any further than just ensuring the locking of resolver-status files (which are a sort of proto-metadata for the resolver to use that helps track unfound artifacts and adhere to an updateInterval for those), we'll need to modify the various Wagons, I think. Since these are file-oriented, they manage opening, writing to, and closing the files internally. One thing I might be able to do is try to manage a file lock on the main artifact file instead of the temp file that the wagon is using as its destination (which may not exist, so that could complicate things even further). Yet another approach might involve creating a lockfile, and managing a lock on that...

I'll look into a couple of these strategies. I don't think it's optimal to bury the locking logic in the various wagons, as we're likely to run into concurrency problems again and again with each marginal wagon implementation.

**Comment by jira-importer** [comment]
**[Christian E Gruber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cgruber)** commented

Frankly, one could mitigate this substantially by downloading artifacts into temporary files, then once the artifact is down with all its pieces, moving the temporary files into the permanent location.  There is still a race condition but it's down to the order of magnitude of milliseconds during the move operation, rather than the many seconds during downloading.

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

This is also discussed in http://docs.codehaus.org/display/MAVEN/Local+repository+separation, along with MNG-3655

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

see also linked issues for other use cases

**Comment by jira-importer** [comment]
**[Paul Benedict](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ID134349)** commented

Giving Maven an exclusive lock on an artifact's local metadata should be a sure way of ensuring this.

**Comment by jira-importer** [comment]
**[John Lonergan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=johnlon)** commented

This issue really impacts our ability to run concurrent  continuous integration builds making out CI take several times longer to cycle than we'd like. We have compensated by having each CI task have it's own private local repo but this massively expands the disk requirements and has the affect of slowing everything down due to all the duplicate file downloading going on. But the really painful thing is that when we have separate local repo's the various threads cannot share the snapshots via the local repo.

Please try to get a fix in place for this.

**Comment by jira-importer** [comment]
**[John Lonergan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=johnlon)** commented

Or possibly someone could advise where the piece of code that maintains the local repo lies and I'll have look.

Things that hurt seem mostly to be that the repo metadata gets corrupted.

**Comment by jira-importer** [comment]
**[Terence Mill](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cforce)** commented

This issue hits us most on a Jenkins buidl server, where all Maven builds use same repo folder. Having many dupes of local repos will use a lot of local hard disk space. Maybe there is a Maven Jenkins Plugin for Jenkins which coordinates the Concurrent access?

**Comment by jira-importer** [comment]
**[Eric Pabst](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=epabst)** commented

It's very frustrating that there is no apparent progress on this.  It was reported years ago and is still hurting productivity significantly.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

You can drop this implementation into your ${M2_HOME}/lib/ext:

http://search.maven.org/#artifactdetails%7Cio.tesla%7Ctesla-concurrent-localrepo%7C0.0.2%7Cjar

I will be integrating this with the Aether version once Maven core is updated to Aether 0.9.0.M1

**Comment by jira-importer** [comment]
**[luke call](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lcall)** commented

Thanks Jason!!

**Comment by jira-importer** [comment]
**[Adam Downer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mungler)** commented

I have read that the aether that's used in maven 3.0.3 is 1.11. Is that correct. Does this mean we could have this fix?

**Comment by jira-importer** [comment]
**[Marc Guenther](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=marc.guenther)** commented

Is this still something we need to do manually, or has this been incorporated into an official maven release?

Can someone confirm that this actually works, or does this have other side effects that prevent this from being released?

**Comment by jira-importer** [comment]
**[Marc Guenther](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=marc.guenther)** commented

Is this still needed for Maven 3.0.5 or 3.1.0?

Also I found a newer version of tesla-concurrent-localrepo (0.0.4), is that the one we should use?

**Comment by jira-importer** [comment]
**[Hannes Kogler](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ntshko)** commented

this is really very time consuming and makes Maven unusable for productive build environments after all.

Is it fixed in Maven 3.1.1?

**Comment by jira-importer** [comment]
**[Jacek Tomaka](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=j-cztery)** commented

Hi, Any news on that? The stable currently is 3.3.3 but I could not find information weather it has been fixed... Thanks.

**Comment by jira-importer** [comment]
**[Jacek Tomaka](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=j-cztery)** commented

Ok, Jason answered but it is not apparent that it was the answer. I guess the answer was: out of the box maven uses local repository implementation that is not thread/executions safe but there is replacement implementation so that local repository can be made thread safe: http://takari.io/book/30-team-maven.html#concurrent-safe-local-repository

Hope this helps other people hitting the wall with their heads.

Thanks Jason and Takari!!!

**Comment by jira-importer** [comment]
**[Michal Domagala](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michaldo)** commented

[J-cztery](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=J-cztery) Local repository may be corrupted by Maven embedded with Eclipse etc. Takari extension is excellent but it is workaround only

**Comment by jira-importer** [comment]
**[Kevin Cross](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kevinhcross)** commented

Are there any plans to merge the Takari fix for this issue into the standard Maven release?

I believe this issue may be incorrectly classified as an improvement when it should be marked as a bug.

When the issue was created it may have been an improvement as CI was less of a common practice but who would be using Maven now in an enterprise environment and not doing CI?

There are work arounds for CI systems but if you don't know about this issue you will not put them in place which will lead to broken builds due to surprising behaviour by the build tool.

**Comment by jira-importer** [comment]
**[Erik Håkansson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erik.hakansson)** commented

This issue plagues a lot of our builds. A lock file with a PID seems like it should solve the issue.

**Comment by jira-importer** [comment]
**[Mike K](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=freshmike)** commented

+1

P.S takari page is down.

**Comment by jira-importer** [comment]
**[Erik Håkansson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erik.hakansson)** commented

The PR mentioned above is for the File wagon in maven-wagon, and is only tested isolated in there.

Since it's pretty much impossible to reliably reproduce this with a real maven, I had to be satisfied with that.

**Comment by jira-importer** [comment]
**[Zakharov Denis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zakharovdi)** commented

Can someone answer on `@erikhakansson` questions or merge PR? We have a lot of build failed due to this issue

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Zakharov Denis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zakharovdi) Have you read this?

> http://takari.io/book/30-team-maven.html#concurrent-safe-local-repository is a Takari Extension for Apache Maven that introduces a concurrent-safe local-repository.

**Comment by jira-importer** [comment]
**[Zakharov Denis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zakharovdi)** commented

I'm not 100% sure but i have put Takari Extension into lib/ext folder and continue to get errors like this yet:
15:11:13.552 [INFO] --------------------------------[ war ]---------------------------------
[|http://fobo-build01.corp.mvideo.ru:8082/jenkins/blue/organizations/jenkins/Build%2FBuildWincash/detail/release%2F20.0.0/72/pipeline/#step-424-log-62]15:11:13.617 [WARNING] The POM for ******:********:jar:20.0.0 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
 
It means that file was changed from another maven process.
 
Mb something changed for years but it doesn't work for me with maven 3.6.1
 
It will be very appreciate if you add concurrent-safe access to local repository in maven core functionality

**Comment by jira-importer** [comment]
**[Gael SELIG](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arkel)** commented

I confirm that it's not working with Takari Extension on maven 3.x, it looks like an abandoned project unfortunately :(

**Comment by jira-importer** [comment]
**[Philippe GRANET](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=phgr100x)** commented

I just found this (not tested): https://github.com/murkaje/robust-maven-cache

**Comment by jira-importer** [comment]
**[Philippe GRANET](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=phgr100x)** commented

See also MRESOLVER-123 for concurrency issues

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

As mentioned in at least one Takari Extension ticket: it is broken, don't use it. MRESOLVER-123 cannot, unfortunately, solve this because it takes care of a single JVM only. My idea for this one is to use a Redis-based ReentrantReadWriteLock.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

MRESOLVER-131 introduces a fully new, multi-JVM concurrent approach based on Redis. Try it from here: https://github.com/apache/maven-resolver/tree/MRESOLVER-131. Read it from here: https://maven.apache.org/resolver-archives/resolver-LATEST/maven-resolver-synccontext-redisson/index.html

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Use any of the two new sync contexts to solve this concurrency issue.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Maven Resolver 1.6.2 will be available with Maven 3.8.2 so this has been addressed in 3.8.x too.

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)

Using redis lock can solve the issue for sure.
Is there any concern that you guys don't use java's FileLock to address this issue? I think FileLock should be enough safe as the maven repo would be only accessed by mvn.  We don't care about other programs, especially for a case in Jenkins.

**Comment by jira-importer** [comment]
**[Anatoly Kupriyanov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kan.izh)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Have you considered to use non-locking approach? I suggest it could be done by renames. I.e. concurrent instances should download file into an unique temporary locations. When download is complete, instance just renames temporary file into a target. And as a rename is atomic, there are no partially downloaded files would be visible by concurrent instances.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Anatoly Kupriyanov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kan.izh), this is not possible. The synchronization does not have access to such low level information. But apart from that, this layman's approach could work.

**Comment by jira-importer** [comment]
**[Daymon Rogers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=darogers)** commented

does anyone know if the redisson locks should work with Tycho/Maven builds?  I see a comment [here](https://svn.apache.org/repos/asf/maven/website/components/resolver-archives/resolver-1.6.3/maven-resolver-synccontext-redisson/index.html) that says: 
* It only works when Sisu DI is used and not the bundled `AetherModule` or `ServiceLocator` (Maven uses Sisu dependency injection).

I don't know enough about the detailed implementation of Maven or Tycho to know where this would apply.

Also, are there any examples available for how to enable the various locking options in my build?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I don't know anything about tycho, but if it runs Maven it will work. There is a Maven branch reconciles Maven 3.8.x and aResolver 1.7.x. I have also pushed new documentation on locks to master.

**Comment by jira-importer** [comment]
**[Stanislav Spiridonov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=foal)** commented

Hi, what a latest status of it? We still have some time corrupted metadata on concurrent builds. Is it better with Maven 3.9? Maybe there can be imlement some workaroud - e.g. reread corrupted data from remote repository, not just report an ERROR? 

```java
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.1.1:install (default-install) on project persistence-model: Failed to install metadata group.id:persistence-model/maven-metadata.xml: Could not parse metadata /opt/mvncache/group/id/persistence-model/maven-metadata-local.xml: in epilog non whitespace content is not allowed but got n (position: END_TAG seen ...</metadata>\nn... @13:2) -> [Help 1] 
```

**Comment by jira-importer** [comment]
**[Henry Kuijpers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henry%20kuijpers)** commented

[Stanislav Spiridonov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=foal) Did your issues get solved?

We also have corruption every now and then. Simple solution is to delete the entire local cache (or the specific corrupted dependencies).

But we feel this should be solveable.

**Comment by jira-importer** [comment]
**[Henry Kuijpers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henry%20kuijpers)** commented

```java
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.1.2:install (default-install) on project parent: Failed to install metadata xxx:parent/maven-metadata.xml: Could not parse metadata /xxx/maven-metadata-local.xml: in epilog non whitespace content is not allowed but got . (position: END_TAG seen ...</metadata>\n.... @12:2)  -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.1.2:install (default-install) on project parent: Failed to install metadata xxx:parent/maven-metadata.xml: Could not parse metadata /xxx/maven-metadata-local.xml: in epilog non whitespace content is not allowed but got . (position: END_TAG seen ...</metadata>\n.... @12:2) 
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:333)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:904)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:281)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:204)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to install metadata xxx:parent/maven-metadata.xml: Could not parse metadata /xxx/maven-metadata-local.xml: in epilog non whitespace content is not allowed but got . (position: END_TAG seen ...</metadata>\n.... @12:2) 
    at org.apache.maven.plugins.install.InstallMojo.installProject (InstallMojo.java:179)
    at org.apache.maven.plugins.install.InstallMojo.execute (InstallMojo.java:132)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:904)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:281)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:204)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
Caused by: org.eclipse.aether.installation.InstallationException: Failed to install metadata xxx:parent/maven-metadata.xml: Could not parse metadata /home/xxx/maven-metadata-local.xml: in epilog non whitespace content is not allowed but got . (position: END_TAG seen ...</metadata>\n.... @12:2) 
    at org.eclipse.aether.internal.impl.DefaultInstaller.install (DefaultInstaller.java:302)
    at org.eclipse.aether.internal.impl.DefaultInstaller.install (DefaultInstaller.java:180)
    at org.eclipse.aether.internal.impl.DefaultInstaller.install (DefaultInstaller.java:136)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.install (DefaultRepositorySystem.java:385)
    at org.apache.maven.plugins.install.InstallMojo.installProject (InstallMojo.java:177)
    at org.apache.maven.plugins.install.InstallMojo.execute (InstallMojo.java:132)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:904)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:281)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:204)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
Caused by: org.eclipse.aether.RepositoryException: Could not parse metadata /home/xxx/maven-metadata-local.xml: in epilog non whitespace content is not allowed but got . (position: END_TAG seen ...</metadata>\n.... @12:2) 
    at org.apache.maven.repository.internal.MavenMetadata.read (MavenMetadata.java:98)
    at org.apache.maven.repository.internal.MavenMetadata.merge (MavenMetadata.java:72)
    at org.eclipse.aether.internal.impl.DefaultInstaller.install (DefaultInstaller.java:291)
    at org.eclipse.aether.internal.impl.DefaultInstaller.install (DefaultInstaller.java:180)
    at org.eclipse.aether.internal.impl.DefaultInstaller.install (DefaultInstaller.java:136)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.install (DefaultRepositorySystem.java:385)
    at org.apache.maven.plugins.install.InstallMojo.installProject (InstallMojo.java:177)
    at org.apache.maven.plugins.install.InstallMojo.execute (InstallMojo.java:132)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:904)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:281)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:204)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314)
Caused by: org.codehaus.plexus.util.xml.pull.XmlPullParserException: in epilog non whitespace content is not allowed but got . (position: END_TAG seen ...</metadata>\n.... @12:2) 
    at org.codehaus.plexus.util.xml.pull.MXParser.parseEpilog (MXParser.java:2090)
    at org.codehaus.plexus.util.xml.pull.MXParser.nextImpl (MXParser.java:1753)
    at org.codehaus.plexus.util.xml.pull.MXParser.next (MXParser.java:1375)
    at org.apache.maven.artifact.repository.metadata.io.xpp3.MetadataXpp3Reader.read (MetadataXpp3Reader.java:589)
    at org.apache.maven.artifact.repository.metadata.io.xpp3.MetadataXpp3Reader.read (MetadataXpp3Reader.java:616)
    at org.apache.maven.repository.internal.MavenMetadata.read (MavenMetadata.java:94)
    at org.apache.maven.repository.internal.MavenMetadata.merge (MavenMetadata.java:72)
    at org.eclipse.aether.internal.impl.DefaultInstaller.install (DefaultInstaller.java:291)
    at org.eclipse.aether.internal.impl.DefaultInstaller.install (DefaultInstaller.java:180)
    at org.eclipse.aether.internal.impl.DefaultInstaller.install (DefaultInstaller.java:136)
    at org.eclipse.aether.internal.impl.DefaultRepositorySystem.install (DefaultRepositorySystem.java:385)
    at org.apache.maven.plugins.install.InstallMojo.installProject (InstallMojo.java:177)
    at org.apache.maven.plugins.install.InstallMojo.execute (InstallMojo.java:132)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:126)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:328)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:316)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:212)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:174)
    at org.apache.maven.lifecycle.internal.MojoExecutor.access$000 (MojoExecutor.java:75)
    at org.apache.maven.lifecycle.internal.MojoExecutor$1.run (MojoExecutor.java:162)
    at org.apache.maven.plugin.DefaultMojosExecutionStrategy.execute (DefaultMojosExecutionStrategy.java:39)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:105)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:73)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:53)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:118)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:261)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:173)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:101)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:904)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:281)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:204)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:314) 
```

**Comment by jira-importer** [comment]
**[Henry Kuijpers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henry%20kuijpers)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Any suggestions or hints on how to tackle this?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Which Maven version?

**Comment by jira-importer** [comment]
**[Henry Kuijpers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henry%20kuijpers)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) version 3.8.6

I just had this issue again, this is what the maven-metadata-local.xml looked like (last time I was unable to capture its contents):

```java
<?xml version="1.0" encoding="UTF-8"?>
<metadata>
  <groupId>xx.xxxxxxxxxxx.xxxx</groupId>
  <artifactId>xxxx-xxxxxx</artifactId>
  <versioning>
    <versions>
      <version>9.30.0-SNAPSHOT</version>
    </versions>
    <lastUpdated>20241114125204</lastUpdated>
  </versioning>
</metadata>
.0.integration-SNAPSHOT</version>
      <version>9.25.0-SNAPSHOT</version>
      <version>9.24.0</version>
      <version>9.25.0.integration-SNAPSHOT</version>
      <version>9.24.1</version>
      <version>9.24.2</version>
      <version>9.24.3</version>
      <version>9.24.4</version>
      <version>9.24.5</version>
      <version>9.24.6</version>
      <version>9.24.7</version>
      <version>9.24.8</version>
      <version>9.26.0-SNAPSHOT</version>
      <version>9.26.0.integration-SNAPSHOT</version>
      <version>9.25.0</version>
      <version>9.25.1</version>
      <version>9.27.0-SNAPSHOT</version>
      <version>9.26.0</version>
      <version>9.27.0.integration-SNAPSHOT</version>
      <version>9.28.0-SNAPSHOT</version>
      <version>9.27.0</version>
      <version>9.27.1</version>
      <version>9.27.1.integration-</version>
      <version>9.28.0.integration-SNAPSHOT</version>
      <version>9.27.2</version>
      <version>9.28.0</version>
      <version>9.29.0-SNAPSHOT</version>
      <version>9.29.0.integration-SNAPSHOT</version>
      <version>9.30.0-SNAPSHOT</version>
      <version>9.30.0.integration-SNAPSHOT</version>
      <version>9.29.0</version>
    </versions>
    <lastUpdated>20241114125202</lastUpdated>
  </versioning>
</metadata>
</metadata> 
```

Notice how the document is "correct", but, the old/original XML document follows (except for the first bytes, which are the new document, which stops earlier).

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Henry Kuijpers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=Henry%20Kuijpers), can you switch to Maven 3.9.x? The concurrency support is much better there.

**Comment by jira-importer** [comment]
**[Henry Kuijpers](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=henry%20kuijpers)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)  Currently running on Maven 3.9.9 and we're having this issue again today.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Create a new issue with the exact outcome and conditions.

**Comment by jira-importer** [comment]
**[liuchangcheng](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=JIRAUSER296009)** commented

is there a new issue about this problem,  we are facing this problem with 3.8.8.
